[APM] Make sure stacked area charts handle no data points incorrectly

dgieselaar opened the pull request:
Closes #40351.

before:
![image](https://user-images.githubusercontent.com/352732/60665511-f07aab00-9e64-11e9-8ae6-612ea0f15230.png)

after:
![image](https://user-images.githubusercontent.com/352732/60665429-b7423b00-9e64-11e9-99c4-7a444b98c7c3.png)

As you can see, data points without any adjacent data do not get drawn. The tooltips still work though:
![image](https://user-images.githubusercontent.com/352732/60665596-29b31b00-9e65-11e9-9437-3654fef1cdd0.png)

dgieselaar committed:
[APM] Make sure stacked area charts handle no data points incorrectly

Closes #40351.

formgeist reviewed (approved):
Looks good to me 👍

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/TransactionBreakdown/TransactionBreakdownGraph/index.tsx:
```diff
@@ -30,19 +31,26 @@ const TransactionBreakdownGraph: React.FC<Props> = props => {
+                ...value,
+                y: undefined
+              };
+        }),
```
I'm struggling a bit following the flow due to the many transformations this takes. It seems this is the three places where breakdown graphs are being transformed:

https://github.com/elastic/kibana/blob/7ad2c947b3d380a967fc1fbea33dec2bd5297152/x-pack/legacy/plugins/apm/server/lib/transactions/breakdown/index.ts#L189-L192

https://github.com/elastic/kibana/blob/7ad2c947b3d380a967fc1fbea33dec2bd5297152/x-pack/legacy/plugins/apm/public/components/shared/TransactionBreakdown/index.tsx#L87-L100

https://github.com/elastic/kibana/blob/7ad2c947b3d380a967fc1fbea33dec2bd5297152/x-pack/legacy/plugins/apm/public/components/shared/TransactionBreakdown/TransactionBreakdownGraph/index.tsx#L29-L38

Would it be possible for you to combine all (or most) of the transformations and do it on the backend?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5787/)

sqren commented:
The deeper we go into the rabbit hole, the more I think we should drop the outline (the additional linechart series) and just use a regular area chart. That would fix all these issues.
https://codesandbox.io/s/6kv44

formgeist commented:
@sqren @dgieselaar I think if that means the approach and style we currently have in our Metrics tab for Memory charts, I'm happy with that compromise. We'll eventually switch to use Elastic Charts, so no need to make complex work-arounds for a styling issue.

<img width="1624" alt="Screenshot 2019-07-04 at 15 18 32" src="https://user-images.githubusercontent.com/4104278/60671599-b9f85c80-9e73-11e9-9932-a52d7e8c821c.png">

sqren commented:
@formgeist yes, that would be the same as the metric charts.

formgeist commented:
Then all good with me to change it and remove the lines.

dgieselaar commented:
As discussed in Slack, we're going to try to keep the style as we have right now if we can do it in a sane way (by using the `cluster` prop).

dgieselaar committed:
Refactor CustomPlot/TransactionCharts

- [x] Keep API calls in a separate file from the useTransactionBreakdown hook
- [x] Bail out of render function early when `receivedDataDuringLifetime` is false
- [x] Use `cluster` prop to draw lines over area instead of creating a new plot
- [x] can lineseries be stacked without creating a new plot?
- [x] Move functionality of SyncChartGroup to ChartsTimeContext (renamed to ChartsSyncContext)
- [x] Only determine noHits inside of CustomPlot instead of using hits.total
- [x] Remove use of getEmptySeries except for testing
- [x] Send TimeSeries from the API instead of mapping data inside of the component
- [x] Use default data in useTransactionBreakdown to prevent is-null checks
- [x] Remove usage of start, end in CustomPlot

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/TransactionBreakdown/TransactionBreakdownGraph/index.tsx:
```diff
@@ -28,33 +25,39 @@ const TransactionBreakdownGraph: React.FC<Props> = props => {
-        data: timeseriesConfig.values,
-        type: 'area',
-        hideLegend: true
+        ...timeseriesConfig,
```
This should be the only transformation left client-side. JSON does not support `undefined` as a value so we cannot transform that server-side.

sqren replied:
Is this transformation still necessary after you added the `isNaN(d.y)` check to `getNull`?
Maybe there are other things at play here but this codesandbox at least seems to work fine with both `{ x: 1 }` and `{ x: 1, y: undefined}`

https://codesandbox.io/s/6kv44

dgieselaar replied:
Yeah, the chart got really weird when I tried that. Presumably because we use a lot of `'y' in coord` checks. I'll try it again though, maybe it was something else that is now fixed.

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/CustomPlot/plotUtils.js:
```diff
@@ -80,7 +80,7 @@ export function getPlotValues(
     XY_HEIGHT: height || XY_HEIGHT,
     XY_WIDTH: width,
-    ...(stacked ? { stackBy: 'y' } : {})
+    stackBy
```
unfortunately as soon as this value is set, weird things happen (with both 'x' and 'y' as a value).

dgieselaar commented:
@sqren everything that was feasible has been addressed I think, LMK if I missed anything:

- [x] Keep API calls in a separate file from the useTransactionBreakdown hook
- [x] Bail out of render function early when `receivedDataDuringLifetime` is false
- [x] Use `cluster` prop to draw lines over area instead of creating a new plot
- [x] Move functionality of SyncChartGroup to ChartsTimeContext (renamed to ChartsSyncContext)
- [x] Only determine noHits inside of CustomPlot instead of using hits.total
- [x] Remove use of getEmptySeries except for testing
- [x] Send TimeSeries from the API instead of mapping data inside of the component
- [x] Use default data in useTransactionBreakdown to prevent is-null checks
- [x] Remove usage of start, end in CustomPlot

sqren commented on x-pack/legacy/plugins/apm/public/components/app/ServiceDetails/MetricsChart.tsx:
```diff
@@ -8,24 +8,23 @@ import React from 'react';
 import { GenericMetricsChart } from '../../../../server/lib/metrics/transform_metrics_chart';
 // @ts-ignore
 import CustomPlot from '../../shared/charts/CustomPlot';
-import { HoverXHandlers } from '../../shared/charts/SyncChartGroup';
```
Thanks for combining this with `useChartsSync` 👍

sqren commented on x-pack/legacy/plugins/apm/public/components/app/ServiceDetails/MetricsChart.tsx:
```diff
@@ -43,8 +42,7 @@ export function MetricsChart({ start, end, chart, hoverXHandlers }: Props) {
       </EuiTitle>
       <CustomPlot
         {...hoverXHandlers}
-        noHits={noHits}
```
Thanks for getting rid of `noHits` up until this point

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/CustomPlot/index.js:
```diff
@@ -170,15 +149,10 @@ export class InnerCustomPlot extends PureComponent {
+    const noHits =
+      flatten(enabledSeries.map(s => s.data)).filter(coord =>
+        isValidCoordinateValue(coord.y)
+      ).length === 0;
```
nit: use `some` instead of `filter` to return a boolean (it's faster too):

```js
const noHits = !flatten(enabledSeries.map(s => s.data))
  .some(coord => isValidCoordinateValue(coord.y))
```

or `every` depending on your prefs:
```js
const noHits = flatten(enabledSeries.map(s => s.data))
  .every(coord => !isValidCoordinateValue(coord.y))
```

sqren commented on x-pack/legacy/plugins/apm/public/components/app/ServiceDetails/ServiceMetrics.tsx:
```diff
@@ -21,25 +21,17 @@ export function ServiceMetrics({ urlParams, agentName }: ServiceMetricsProps) {
+            </EuiPanel>
+          </EuiFlexItem>
+        ))}
+      </EuiFlexGrid>
```
Yay! No more HOCs!

sqren replied:
... or rather: render props. But same same. Both legacy now we have hooks and context :)

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/TransactionBreakdown/TransactionBreakdownGraph/index.tsx:
```diff
@@ -7,17 +7,14 @@
-    color: string;
-    values: Array<{ x: number; y: number | null }>;
-  }>;
+  timeseries: TimeSeries[];
```
👌

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/CustomPlot/StaticPlot.js:
```diff
@@ -21,8 +21,13 @@ import { rgba } from 'polished';
+import { isValidCoordinateValue } from '../../../../utils/isValidCoordinateValue';
+
+// undefined values are converted by react-vis into NaN when stacking
+const getNull = d => isValidCoordinateValue(d.y) && !isNaN(d.y);
```
Probably should add a link to the react-vis issue here

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/CustomPlot/StaticPlot.js:
```diff
@@ -48,17 +53,43 @@ class StaticPlot extends PureComponent {
+            stack={true}
+            cluster="line"
+          />
+        ];
```
Yay! A new chart type! <3

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/CustomPlot/StaticPlot.js:
```diff
@@ -122,7 +143,19 @@ class StaticPlot extends PureComponent {
+              }}
+            />,
+            this.getVisSeries(series, plotValues)
+          ]
```
This will remove the y-axis and the horizontal grid lines when there is no data. Probably not a big issue but why? 
Isn't it nice to keep these thing around so the chart doesn't "jump"?
<img width="1214" alt="Screenshot 2019-07-05 at 14 13 59" src="https://user-images.githubusercontent.com/209966/60721778-4ddb2e80-9f2f-11e9-8d9a-38e950b3adcc.png">

formgeist replied:
agreed, that makes the charts behave differently, at least the transactions duration and requests per minute ones. 
<img width="1632" alt="Screenshot 2019-07-05 at 14 28 07" src="https://user-images.githubusercontent.com/4104278/60722432-2ab17e80-9f31-11e9-909f-bd26cd47e6fc.png">

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/TransactionLineChart/index.tsx:
```diff
@@ -26,82 +21,37 @@ interface Props {
       yMax={yMax}
       height={height}
       truncateLegends={truncateLegends}
+      {...(stacked ? { stackBy: 'y' } : {})}
```
What about:
```js
stackBy: {stacked ? 'y' : undefined}
```

sqren replied:
or perhaps just pass in `stacked` and handle it in plotUtils but with `undefined` as the fallback value.

sqren commented on x-pack/legacy/plugins/apm/public/components/app/ServiceDetails/MetricsChart.tsx:
```diff
@@ -34,7 +33,7 @@ export function MetricsChart({ start, end, chart, hoverXHandlers }: Props) {
   }));
 
-  const noHits = chart.totalHits === 0;
+  const { hoverXHandlers } = useChartsSync();
```
What do you think about skipping the intermediary object. `hoverXHandlers` is also an odd name :D
```js
const { onHover, onMouseLeave, onSelectionEnd } = useChartsSync();
```

Consumers can still use the other approach if they want:
```js
const hoverXHandlers = useChartsSync();
```

sqren commented on x-pack/legacy/plugins/apm/public/hooks/useTransactionBreakdown.ts:
```diff
@@ -6,27 +6,27 @@
+        end,
+        serviceName,
+        transactionName,
+        uiFilters
```
I see what you mean with the extra abstraction and it doesn't even make it more DRY.

sqren commented on x-pack/legacy/plugins/apm/public/selectors/chartSelectors.ts:
```diff
@@ -61,7 +58,6 @@ export function getTransactionCharts(
-    noHits,
     tpmSeries,
     responseTimeSeries
   };
```
I think this file can also be moved to the backend (https://github.com/elastic/kibana/blob/master/x-pack/legacy/plugins/apm/server/lib/transactions/charts/get_timeseries_data/transform.ts) but that's probably out of scope for this PR :)

dgieselaar replied:
Agree on both counts. Would be easier if all endpoints spit out TimeSeries, but also feel like this PR is getting a bit big for a bug fix.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5859/)

dgieselaar committed:
Additional cleanup on aisle five

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6030/)

sqren commented on x-pack/legacy/plugins/apm/public/context/ChartsSyncContext.tsx:
```diff
@@ -0,0 +1,55 @@
+      hoverX: time
+    };
+
+    return { ...hoverXHandlers };
```
Or simply:

```suggestion
    return hoverXHandlers;
```

dgieselaar replied:
Aye, I settled on `syncedChartProps` 🤓

sqren commented on x-pack/legacy/plugins/apm/public/hooks/useTransactionCharts.ts:
```diff
@@ -7,29 +7,32 @@
+  } = useUrlParams();
 
   const { data, error, status } = useFetcher(() => {
     if (serviceName && start && end) {
```
`transactionName` used to be required here. I recall something about the overview chart being loaded when navigating to the details page when `transactionName` was not immediately available (or vice versa) which is way I separated those two. Might not be a problem anymore and then this simplification is good, just double checking.

dgieselaar replied:
I thought about this, and then figured this should be pretty safe as the values come from the URL which is stable. Except... for redirects, I'll make sure that  check that.

sqren commented on x-pack/legacy/plugins/apm/public/selectors/chartSelectors.ts:
```diff
@@ -158,12 +159,18 @@ export function getTpmSeries(
+
+  if (!tpmBuckets.length && avg.length) {
+    return getEmptySeries(avg[0].x, avg[avg.length - 1].x);
+  }
```
I'm not sure I understand this logic :)
As I read it an empty state series is returned if there are no tpm buckets but there is data for avg series. 
What happens when there is no data for either tpm buckets or avg series?

And how come `getEmptySeries` is necessary here, when it's not necessary for the other charts?

dgieselaar replied:
I should add a comment here. The aggs in `apmTimeSeries` (`avg`,`p95`,`p99`) always return date buckets with null values, but `tpmBuckets` is from a term aggregation, meaning we get no date buckets, because we have no terms. We could send an empty series from the backend in those cases as well, maybe that is more obvious?

sqren replied:
The empty states still look off. The tick marks on the y-axis are missing (halfway and bottom)

<img width="2402" alt="Screenshot 2019-07-08 at 11 56 11" src="https://user-images.githubusercontent.com/209966/60801532-6fc6f200-a177-11e9-80a1-67b57ad4ceec.png">

dgieselaar replied:
@sqren hmm, yeah, I hoped the `yTickValues` prop would fix this, but it's probably because there are only `null` coords. `getEmptySeries` creates a timeseries with coordinates of `y: 1`. We could fix it by reinstating `getEmptySeries` inside of CustomPlot again. Do you think it's worth it?

sqren replied:
> Do you think it's worth it?

Hmm.. I think so. There are (minor) regressions to the design and I'm seeing this error:

<img width="817" alt="Screenshot 2019-07-08 at 13 50 52" src="https://user-images.githubusercontent.com/209966/60808001-8a08cc00-a187-11e9-8835-269d2bd5ba78.png">

dgieselaar replied:
@sqren yeah, the alternative would be to hide the gridlines and y axis in that case, like how it was in the initial implementation of this PR I think. I'll put back the empty series w/ a value of 1 (but it feels kinda gross).

dgieselaar replied:
As discussed via Zoom, we decided to drop the YAxis and HorizontalGridLines if there's no data. The cost of the alternative is Too Damn High™, at least in terms of maintainability.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/CustomPlot/plotUtils.js:
```diff
@@ -71,7 +71,9 @@ export function getPlotValues(
-  const yTickValues = [0, yMaxNice / 2, yMaxNice];
+  const yTickValues = !isNaN(yMaxNice)
+    ? [0, yMaxNice / 2, yMaxNice]
+    : [0, 0.5, 1];
```
This change doesn't look right. If `yMaxNice` is `NaN` we'll have other problems too.

dgieselaar replied:
it's no longer necessary if we pass an empty series with `1` as an 'empty' value rather than `null`. I do think we're adding hack on hack here, but maybe I have not considered all possible requirements yet.

sqren replied:
How will the new `noHits` check work when you are passing in `1`?

dgieselaar replied:
we only use 1 internally (in CustomPlot). Let me see how it would look.

dgieselaar committed:
Don't show GridLines or YAxis component if there is no data

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6160/)

Conky5 commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6199/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6297/)

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/index.tsx:
```diff
@@ -53,22 +54,18 @@ const msTimeUnitLabel = i18n.translate(
-  public getResponseTimeTickFormatter = (t: number | null) => {
-    return this.props.charts.noHits ? `- ${msTimeUnitLabel}` : asMillis(t);
+  public getResponseTimeTickFormatter = (t: number | null | undefined) => {
+    return isValidCoordinateValue(t) ? asMillis(t) : `- ${msTimeUnitLabel}`;
```
```suggestion
    return asMillis(t);
```

I don't see how react-vis will return an invalid coordinate (that would fail earlier afaict)

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/index.tsx:
```diff
@@ -53,22 +54,18 @@ const msTimeUnitLabel = i18n.translate(
-    return this.props.charts.noHits || !p
-      ? `- ${msTimeUnitLabel}`
-      : asMillis(p.y);
+    return isValidCoordinateValue(p.y) ? asMillis(p.y) : `- ${msTimeUnitLabel}`;
```
```suggestion
    return asMillis(p.y);
```

as above

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/index.tsx:
```diff
@@ -53,22 +54,18 @@ const msTimeUnitLabel = i18n.translate(
-    return charts.noHits || t === null
-      ? `- ${unit}`
-      : `${asInteger(t)} ${unit}`;
+    return isValidCoordinateValue(t) ? `${asInteger(t)} ${unit}` : `- ${unit}`;
```
```suggestion
    return `${asInteger(t)} ${unit}`;
```

as above

dgieselaar committed:
Make formatters less defensive

dgieselaar commented:
@sqren Looks like your suggested change does not pass type checking. I've pushed an alternative, LMK what you think: https://github.com/elastic/kibana/pull/40353/commits/d9e0d4cd9a95b11cd31da3c7d6d2720084feb680

sqren commented:
@dgieselaar lgtm! 👍

dgieselaar committed:
Fix loading state for TransactionCharts

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6354/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6412/)

dgieselaar committed:
Update translation files

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6525/)
